### PR TITLE
Enable NCHW/NHWC and NCDHW/NDHWC layout  in batch norm driver command

### PR DESCRIPTION
**Pull request by bghimireamd**
- Enable NCHW/NHWC layout from driver command for batch norm. 
- Moved `GpumemTensor` to `driver/driver.hpp`
- Stopped using old and slow `miopenBNFwdTrainPerActivationRunHost`, `miopenBNFwdTrainSpatialRunHost`, `miopenBNFwdInferPerActivationRunHost`, `miopenBNFwdInferSpatialRunHost`, `miopenBNBwdPerActivationRunHost` and `miopenBNBwdSpatialRunHost` since they only handled NCHW layout. Started using layout agnostic batch norm host function from `test/fusionHost.hpp`

eg: 

`./bin/MIOpenDriver bnorm -n 64 -c 1024 -H 14 -W 14 -m 1 --forw 1 -b 0 -s 0 -r 1`  _// defaults to NCHW_

`./bin/MIOpenDriver bnorm -n 64 -c 1024 -H 14 -W 14 -m 1 --forw 1 -b 0 -s 0 -r 1 -L NCHW`
`./bin/MIOpenDriver bnorm -n 64 -c 1024 -H 14 -W 14 -m 1 --forw 1 -b 0 -s 0 -r 1 -L NHWC`
`./bin/MIOpenDriver bnorm -n 64 -c 1024 -H 14 -W 14 -m 1 --forw 1 -b 0 -s 0 -r 1 --layout NHWC`

`bnorm, bnormfp16, bnormfp16fp32, bnormbfp16fp32,`

3D (we do not have 3D solvers we convert 3D to 2D tensor and feed it to 2D solvers. Once we have 3D batch norm solvers we can remove the conversion from 3D to 2D tensor)

3D example
`./bin/MIOpenDriver bnorm -n 8 -c 8 -H 12 -W 12 -D 12 -m 1 --forw 0 -b 1 -s 1 -r 1 --layout NDHWC`

### CODE REVIEW

**Review by alexandraBara — COMMENTED** [review]
-L/--layout is quite generic. Which layout does it refer to? in/out/fil? 
In convolution we do this: --fil_layout NCHW --in_layout NCHW --out_layout NCHW - which is nice bc its specific. I think BN should follow the same pattern.
Maybe there will be a time where in/fil layout in BN will be different, so we need to keep that in mind

File: driver/bn_driver.hpp [hunk 1]
```diff
@@ -60,19 +63,157 @@
...
+            ///
+            /// However we do not care about precision in gpualloc mode, because validation
+            /// is not used. Therefore, range (0,1] is fine.
+            return gpumemrand::gen_0_1(static_cast<Tgpu*>(GetDevicePtr()), sz);
+        }
+        return dev->ToGPU(q, GetVectorData());
+    }
+
+    template <typename T>
+    status_t AllocOnDevice(stream, context_t ctx, const size_t sz, std::vector<T>&)
```
Comment (CAHEK7) [inline driver/bn_driver.hpp]: That's suspicious:
https://github.com/ROCm/MIOpen/blob/7e161f53d1e6e9b4ecf0fdbf59bb452b4f749c09/driver/conv_driver.hpp#L338
I'm not sure that we want another copy.

File: driver/bn_driver.hpp [hunk 2]
```diff
@@ -60,19 +63,157 @@
...
+    }
+
+    auto GetDevicePtr() -> auto { return dev->GetMem(); }
+};
+//========================
+
 //#define BN_RUNFOR_PROFILER
 
-template <typename Tgpu, typename Tref, typename Tmix = Tgpu>
+template <typename Tgpu, typename Acc, typename Tref, typename Tmix = Tgpu>
```
Comment (CAHEK7) [inline driver/bn_driver.hpp]: It doesn't look like layout related, could you clarify that change?

**Review by CAHEK7 — DISMISSED** [review]
Some minor suggestions

File: driver/bn_driver.hpp (hunk 2 above)
Comment (bghimireamd) [reply driver/bn_driver.hpp]: reverted

File: driver/bn_driver.hpp (hunk 1 above)
Comment (bghimireamd) [reply driver/bn_driver.hpp]: should I bring class `GpumemTensor` from `conv_driver.hpp` into another file `driver_common.hpp` and let both `conv_driver.hpp` and `bn_driver.hpp` use it.

Comment (CAHEK7) [reply driver/bn_driver.hpp]: Yes, and somewhen, in the future, when you will see any raw or even managed gpu pointers in the driver\tests, you can recommend to use that class.

Comment (bghimireamd) [reply driver/bn_driver.hpp]: For now I moved to `driver.hpp`

File: driver/dm_bnorm.cpp [hunk 3]
```diff
@@ -31,7 +31,11 @@ static Driver* makeDriver(const std::string& base_arg)
     if(base_arg == "bnorm")
         return new BatchNormDriver<float, double>();
     if(base_arg == "bnormfp16")
+        return new BatchNormDriver<float16, double, float16>();
+    if(base_arg == "bnormfp16fp32")
```
Comment (alexandraBara) [inline driver/dm_bnorm.cpp:35]: "bnormfp16fp32" ? is this a typo?

**Review by alexandraBara — DISMISSED** [review]

File: driver/driver.hpp [hunk 4]
```diff
@@ -157,6 +160,140 @@ struct GPUMem
 #endif
 };
 
+template <typename Tgpu>
+class GpumemTensor
```
Comment (alexandraBara) [inline driver/driver.hpp:164]: Can this class be exposed in include/miopen? So we can include it in our Fin code? Otherwise, I think we are reinventing the wheel here: https://github.com/ROCm/MIFin/blob/develop/src/include/tensor.hpp

Comment (alexandraBara) [reply driver/driver.hpp:164]: @CAHEK7  thoughts?

File: driver/dm_bnorm.cpp (hunk 3 above)
Comment (bghimireamd) [reply driver/dm_bnorm.cpp:35]: Unfortunately, no. We have two different batch normalization kernels: OCL and CK. Both use different data types for their tensors. See the image below

image

For example, dScale is fp16 in the OCL kernel and fp32 in the CK kernel. Thus, bnormfp16fp32 is used for CK, while bnormfp16 is used for the OCL kernel's tensor initialization. This is an unsatisfactory solution for now, until we can develop a proper final API.

File: driver/driver.hpp (hunk 4 above)
Comment (CAHEK7) [reply driver/driver.hpp:164]: It should be in the next PR. Lets separate unrelated changes.

File: src/driver_arguments.cpp [hunk 5]
```diff
@@ -66,6 +66,10 @@ void BnDataType(std::stringstream& ss, const miopen::TensorDescriptor& desc)
     {
         ss << "bnormfp16";
     }
+    if(desc.GetType() == miopenBFloat16)
+    {
+        ss << "bnormbfp16";
+    }
```
Comment (BrianHarrisonAMD) [inline src/driver_arguments.cpp:76]: Not sure we have enough details here to know, but do we need to handle "bnormfp16fp32" here as well?

And should this be "bnormbfp16fp32" instead?

Comment (bghimireamd) [reply src/driver_arguments.cpp:76]: you are correct! There is not enough details to find out where the run was "bnormfp16" or "bnormfp16fp32" so I defaulted to "bnormfp16"

File: include/miopen/miopen.h [hunk 6]
```diff
@@ -2738,6 +2738,28 @@ miopenBatchNormalizationForwardTraining(miopenHandle_t handle,
                                         void* resultSaveMean,
                                         void* resultSaveInvVariance);
 
+MIOPEN_EXPORT miopenStatus_t
+miopenBatchNormalizationForwardTraining_V2(miopenHandle_t handle,
```
Comment (bghimireamd) [inline include/miopen/miopen.h:2781]: since C does not support function overloading had to create `miopenBatchNormalizationForwardTraining_V2` for `miopenBatchNormalizationForwardTraining`

**Comment by junliume** [comment]
@bghimireamd the FIN submodule needs to be updated from the CI results. Thanks!

**Review by alexandraBara — APPROVED** [review]
LGTM

File: driver/bn_driver.hpp [hunk 7]
```diff
@@ -317,13 +344,92 @@ std::vector<int> BatchNormDriver<Tgpu, Tref, Tmix>::GetInputTensorLengthsFromCmd
...
+template <typename Tgpu, typename Tref, typename Tmix>
+int BatchNormDriver<Tgpu, Tref, Tmix>::ChkLayout_ShortName()
+{
+    // check for short name of layout type
+    if(inflags.FindShortName("layout") == 'L')
+    {
+        // do noting
+        // found valid short names
+        return 0;
+    }
```
Comment (BrianHarrisonAMD) [inline driver/bn_driver.hpp]: Just a nitpick, but surprised this returns an int, and 0 is valid. 
I would expect it to return a bool, and true means valid.

File: driver/bn_driver.hpp [hunk 8]
```diff
@@ -317,13 +344,92 @@ std::vector<int> BatchNormDriver<Tgpu, Tref, Tmix>::GetInputTensorLengthsFromCmd
...
+           (layout_value.compare("NCDHW") == 0) || (layout_value.compare("NDHWC") == 0))
+        {
+            // do nothing,Values are matching as defined in Lib.
+        }
+        else
+        {
+            std::cerr << "Invalid Layout Parameter Value - " << layout_value << std::endl;
+            exit(EXIT_FAILURE);
+        }
+    }
```
Comment (BrianHarrisonAMD) [inline driver/bn_driver.hpp]: Could collapse this down into a single if / else if block to simplify:


```cpp
if((ChkLayout_ShortName()))
{
    std::cerr << " Invalid Layout Short Name = " << ChkLayout_ShortName() << std::endl;
    exit(EXIT_FAILURE);
}
else if((layout_value.compare("NCHW") != 0) && (layout_value.compare("NHWC") != 0) &&
           (layout_value.compare("NCDHW") != 0) && (layout_value.compare("NDHWC") != 0)
{
     std::cerr << "Invalid Layout Parameter Value - " << layout_value << std::endl;
     exit(EXIT_FAILURE);
}
```

File: src/batch_norm_api.cpp [hunk 9]
```diff
@@ -88,27 +90,156 @@ miopenBatchNormalizationForwardInference(miopenHandle_t handle,
...
                         bn_mode,
                         xDesc,
                         x,
                         yDesc,
                         y,
-                        bnScaleBiasMeanVarDesc,
+                        scaleDesc,
+                        (BiasDesc == nullptr) ? scaleDesc : BiasDesc,
+                        (estMeanDesc == nullptr) ? scaleDesc : estMeanDesc,
+                        (estVarianceDesc == nullptr) ? scaleDesc : estVarianceDesc,
```
Comment (BrianHarrisonAMD) [inline src/batch_norm_api.cpp]: If we change the original API calls to instead call _V2 with these set, then we could make these descriptors required for the _V2 API, and not need these checks.

It's not a huge deal, but I think doing that would simplify the code a bit.

(If you decide to do this change, then the same also applies for training & backwards)

File: src/solver/batchnorm/backward_ck.cpp [hunk 10]
```diff
@@ -201,7 +201,8 @@ bool BnCKBwdBackward::IsApplicable(
         return false;
     if(bn_problem.GetDirection() != miopen::batchnorm::Direction::Backward)
         return false;
-
+    if(!bn_problem.Is2D())
+        return false;
```
Comment (BrianHarrisonAMD) [inline src/solver/batchnorm/backward_ck.cpp:205]: Should this have always been there?

I would have thought this was needed before these changes as well.

File: test/gtest/bn_bwd.cpp [hunk 11]
```diff
@@ -26,46 +26,123 @@
...
+TEST_P(GPU_BN_V2_BwdNHWC_FP64, BnV2BwdCKDouble) {}
 
-TEST_P(GPU_BNBwd_BFP16, BnBwdCKBFloat16) {}
-TEST_P(GPU_BNBwd_FP64, BnBwdCKDouble) {}
+// // fp16
+// INSTANTIATE_TEST_SUITE_P(Smoke,
+//                          GPU_BN_V1_BwdNCHW_FP16,
+//                          testing::Combine(testing::ValuesIn(NetworkSmall<BNTestCase>()),
+//                                           testing::Values(miopenTensorNCHW),
+//                                           testing::ValuesIn({testBNAPIV1})));
```
Comment (BrianHarrisonAMD) [inline test/gtest/bn_bwd.cpp:103]: Any reason why this one is commented out?

**Review by BrianHarrisonAMD — APPROVED** [review]
LGTM.

Left some suggestions, but don't think they are urgent.

File: driver/bn_driver.hpp (hunk 7 above)
Comment (bghimireamd) [reply driver/bn_driver.hpp]: Yes! I borrowed the code from conv_driver.hpp and did not bothered to change. I will make this `bool` since that makes more sense here.

File: src/batch_norm_api.cpp (hunk 9 above)
Comment (bghimireamd) [reply src/batch_norm_api.cpp]: good call

File: src/solver/batchnorm/backward_ck.cpp (hunk 10 above)
Comment (bghimireamd) [reply src/solver/batchnorm/backward_ck.cpp:205]: yes! current CK batch norm only supports 2D.

File: test/gtest/bn_bwd.cpp (hunk 11 above)
Comment (bghimireamd) [reply test/gtest/bn_bwd.cpp:103]: I commented this since the ocl batch norm solver NCHW is not working. I think we need to fix that but that is separate task. I have commented for now so that we don't forget this issue.

**Review by cderb — APPROVED** [review]

**Review by junliume — APPROVED** [review]
